### PR TITLE
fix: Avoid set_state on readonly objects

### PR DESCRIPTION
**Pull request by mkundu1**

### CODE REVIEW

File: src/ansys/fluent/core/services/datamodel_se.py [hunk 1]
```diff
@@ -88,6 +88,13 @@ def __init__(self, request):
         super().__init__(f"Failed to unsubscribe event: {request}!")
 
 
+class ReadOnlyObjectError(RuntimeError):
+    """Raised when readonly object is mutated."""
```
Comment (PipKat) [inline src/ansys/fluent/core/services/datamodel_se.py:92]:
```suggestion
    """Raised when a read-only object is mutated."""
```

File: src/ansys/fluent/core/services/datamodel_se.py [hunk 2]
```diff
@@ -760,7 +767,20 @@ def fix_state(self) -> None:
...
+
+        Parameters
+        ----------
+        state : Any, optional
+            state
+
+        Raises
+        ------
+        ReadOnlyObjectError
+            If the object is readonly.
```
Comment (PipKat) [inline src/ansys/fluent/core/services/datamodel_se.py:780]:
```suggestion
            If the object is read-only.
```

File: tests/test_datamodel_service.py [hunk 3]
```diff
@@ -472,3 +473,10 @@ def test_command_creation_inside_singleton(new_mesh_session):
     meshing = new_mesh_session
     read_mesh = meshing.meshing.File.ReadMesh.create_instance()
     assert read_mesh.FileName is not None
+
+
+def test_read_ony_set_state(new_mesh_session):
+    meshing = new_mesh_session
+    assert meshing.preferences.MeshingWorkflow.CheckpointingOption.is_read_only()
+    with pytest.raises(ReadOnlyObjectError):
+        meshing.preferences.MeshingWorkflow.CheckpointingOption = "Write into memory"
```
Comment (PipKat) [inline tests/test_datamodel_service.py:482]:
```suggestion
        meshing.preferences.MeshingWorkflow.CheckpointingOption = "Write to memory"
```

Comment (mkundu1) [reply tests/test_datamodel_service.py:482]: The RHS string is coming from Fluent, cannot be changed here.

**Review by prmukherj — APPROVED** [review]

**Review by hpohekar — APPROVED** [review]

File: tests/test_datamodel_service.py [hunk 4]
```diff
@@ -472,3 +473,10 @@ def test_command_creation_inside_singleton(new_mesh_session):
     meshing = new_mesh_session
     read_mesh = meshing.meshing.File.ReadMesh.create_instance()
     assert read_mesh.FileName is not None
+
+
+def test_read_ony_set_state(new_mesh_session):
+    meshing = new_mesh_session
+    assert meshing.preferences.MeshingWorkflow.CheckpointingOption.is_read_only()
```
Comment (seanpearsonuk) [inline tests/test_datamodel_service.py:480]: @mkundu1 could this assertion fail due to arbitrary initial preferences state?

File: tests/test_datamodel_service.py [hunk 5]
```diff
@@ -472,3 +473,10 @@ def test_command_creation_inside_singleton(new_mesh_session):
     meshing = new_mesh_session
     read_mesh = meshing.meshing.File.ReadMesh.create_instance()
     assert read_mesh.FileName is not None
+
+
+def test_read_ony_set_state(new_mesh_session):
+    meshing = new_mesh_session
```
Comment (seanpearsonuk) [inline tests/test_datamodel_service.py:479]:
```suggestion
    meshing = new_mesh_session
    meshing.preferences.MeshingWorkflow.SaveCheckpointFiles = True
```

File: tests/test_datamodel_service.py (hunk 3 above)
Comment (seanpearsonuk) [inline tests/test_datamodel_service.py:482]: @mkundu1 Is meshing.preferences.MeshingWorkflow.set_state({...}) also covered?

File: src/ansys/fluent/core/services/datamodel_se.py [hunk 6]
```diff
@@ -88,6 +88,13 @@ def __init__(self, request):
         super().__init__(f"Failed to unsubscribe event: {request}!")
 
 
+class ReadOnlyObjectError(RuntimeError):
```
Comment (seanpearsonuk) [inline src/ansys/fluent/core/services/datamodel_se.py:91]: For 25R1, we should implement this behaviour in the server.

File: src/ansys/fluent/core/services/datamodel_se.py [hunk 7]
```diff
@@ -760,7 +767,20 @@ def fix_state(self) -> None:
     fixState = fix_state
 
     def set_state(self, state: Optional[Any] = None, **kwargs) -> None:
-        """Set state of the current object."""
+        """Set state of the current object.
```
Comment (seanpearsonuk) [inline src/ansys/fluent/core/services/datamodel_se.py:770]: Another scenario might be via `update_dict()`.

File: src/ansys/fluent/core/services/datamodel_se.py [hunk 8]
```diff
@@ -88,6 +88,13 @@ def __init__(self, request):
         super().__init__(f"Failed to unsubscribe event: {request}!")
 
 
+class ReadOnlyObjectError(RuntimeError):
+    """Raised when a read-only object is mutated."""
```
Comment (seanpearsonuk) [inline src/ansys/fluent/core/services/datamodel_se.py:92]:
```suggestion
    """Raised on an attempt to mutate a read-only object."""
```

File: tests/test_datamodel_service.py (hunk 3 above)
Comment (prmukherj) [reply tests/test_datamodel_service.py:482]: yes.

Comment (seanpearsonuk) [reply tests/test_datamodel_service.py:482]: OK, is it covered by the tests?

**Comment by prmukherj** [comment]
@mkundu1, I have updated the dir call.

File: tests/test_datamodel_service.py (hunk 3 above)
Comment (mkundu1) [reply tests/test_datamodel_service.py:483]: As discussed, we shall handle this at StateEngine level in server.

**Pull request merged**